### PR TITLE
Fixes #24729: Allow using a different password hash algorithm for each local user

### PR DESCRIPTION
**Pull request by amousset**
https://issues.rudder.io/issues/24729

* Makes a new `RudderPasswordEncoder` class that acts as a redirection to the correct encoder based on encoded password format
* Try to encapsulate better the internals of password management inside `RudderPasswordEncoder`
* Add a simple `securityLevel` lever to only allow legacy hashes when necessary
* Add the migration logic on the users file, run at startup and on reload, by using a boolean attribute `unsafe-hashes` to still be able to use legacy hashes

### CODE REVIEW

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by clarktsiory** [comment]
PR rebased

**Comment by clarktsiory** [comment]
PR rebased since the merge of user-management. Since the algorithm is now by user, the API of user-management always return the default digest (BCRYPT), which is displayed as a hint when filling the password in the UI. We will need a PR to replace that UI field, and maybe infer and apply validations now that there are always known constraints.

File: webapp/sources/rudder/rudder-web/src/main/scala/bootstrap/liftweb/checks/migration/CheckUsersFile.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+class CheckUsersFile(migration: UserFileSecurityLevelMigration) extends BootstrapChecks {
+
+  override def description: String = "Check if hash algorithm for user password is a modern one, if not enable unsafe_hashes"
+
+  def allChecks(currentSecurityLevel: SecurityLevel): IOResult[Unit] = {
+    for {
+      _ <- currentSecurityLevel match {
+             case SecurityLevel.Modern => ZIO.unit
+             case SecurityLevel.Legacy => migration.migrateToModern(UserManagementIO.getUserFilePath(migration.file))
```
Comment (amousset) [inline webapp/sources/rudder/rudder-web/src/main/scala/bootstrap/liftweb/checks/migration/CheckUsersFile.scala:58]: I think `migrateToModern` is not a good name as it does the exact opposite, maybe "allowLegacy" or something like this.

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/users/RudderUserDetailsFile.scala [hunk 2]
```diff
@@ -377,46 +493,49 @@ object UserFileProcessing {
...
+          ) *> Unexpected(s"rudder-users configuration file not found at path: '${config.path}'").fail
         }
     }
   }
 
+  def readUserFile(resource: UserFile): IOResult[Elem] = {
+    IOResult.attempt {
+      scala.xml.XML.load(resource.inputStream())
+    }.mapError {
+      // map a SAXParseException to a technical butmore user-friendly but error message
```
Comment (fanf) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/users/RudderUserDetailsFile.scala:515]: too many but in that comment

File: webapp/sources/rudder/rudder-web/src/main/scala/bootstrap/liftweb/checks/migration/CheckUsersFile.scala (hunk 1 above)
Comment (fanf) [reply webapp/sources/rudder/rudder-web/src/main/scala/bootstrap/liftweb/checks/migration/CheckUsersFile.scala:58]: `allowLegacy` seems good

**Review by fanf — COMMENTED** [review]
Seems good. 

Can you had a unit test of : 
- a user file with unsafe algo getting chechekd and migrated ?
- a user file with bcrypt getting checked and not migrated ?

**Comment by clarktsiory** [comment]
PR updated

**Comment by amousset** [comment]
There is now a conflict, rebase required @clarktsiory

**Comment by clarktsiory** [comment]
PR rebased

**Pull request merged**